crates/dbsp/src/trace/layers/column_layer/cursor.rs

```diff
@@ -43,15 +43,33 @@ where
         P: Fn(&K) -> bool,
     {
         unsafe { self.storage.assume_invariants() }
-        self.pos += advance(&self.storage.keys[self.pos..self.bounds.1], predicate);
+        if self.valid() {
+            self.pos += advance(
+                &self.storage.keys[self.pos as usize..self.bounds.1],
+                predicate,
+            ) as isize;
+        }
+    }
+
+    pub fn seek_key_with_reverse<P>(&mut self, predicate: P)
+    where
+        P: Fn(&K) -> bool,
+    {
+        unsafe { self.storage.assume_invariants() }
+        if self.valid() {
+            self.pos -= retreat(
+                &self.storage.keys[self.bounds.0..=self.pos as usize],
+                predicate,
+            ) as isize;
+        }
     }
 
     pub fn current_key(&self) -> &K {
-        &self.storage.keys[self.pos]
+        &self.storage.keys[self.pos as usize]
```

Comment:
perhaps this is paranoia, but could this be checked for overflow?

Reply:
I am guessing that pos is always positive, but there is no subtraction for usize?

Reply:
I should add some comments about it. `pos` is `isize`, so we can represent an invalid cursor that moved beyond the left end of the range as `-1`. `current_key` is not allowed to be called on an invalid cursor. I will add debug checks for this.